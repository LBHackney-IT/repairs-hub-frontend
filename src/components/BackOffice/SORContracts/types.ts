export interface Errors {
    selectedOption?: string
    sourcePropertyReference?: string
    destinationPropertyReference?: string
    workOrderReferences?: string
    contractReference?: string
}

export interface RadioInputOption {
    text: string
    value: string
}