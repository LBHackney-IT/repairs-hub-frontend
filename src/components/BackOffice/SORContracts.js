import Layout from "./Layout"

const SORContracts = () => {

  return (
    <Layout title="Add SOR contracts to properties">

    </Layout>
  )
}

export default SORContracts
