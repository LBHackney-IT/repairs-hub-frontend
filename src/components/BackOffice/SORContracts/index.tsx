import Layout from '../Layout'
import React, { useState } from 'react'
import { TextInput, Button } from '../../Form'
import ControlledRadio from '../Components/ControlledRadio'
import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'
import Spinner from '../../Spinner'
import ErrorMessage from '../../Errors/ErrorMessage'
import SuccessMessage from '../Components/SuccessMessage'

import { Errors, RadioInputOption } from './types'

const radioOptions: Array<RadioInputOption> = [
  {
    text: 'Copy contracts from another property',
    value: 'Copy',
  },
  {
    text: 'Add explicit property and contract references',
    value: 'Add',
  },
]

const SORContracts = () => {
  const [selectedOption, setSelectedOption] = useState(radioOptions[0].value)

  const [sourcePropertyReference, setSourcePropertyReference] = useState('')
  const [
    destinationPropertyReference,
    setDestinationPropertyReference,
  ] = useState<string>('')
  const [contractReference, setContractReference] = useState<string>('')

  const [errors, setErrors] = useState<Errors>({})

  const [loading, setLoading] = useState<boolean>(false)
  const [requestError, setRequestError] = useState(null)
  const [formSuccess, setFormSuccess] = useState<boolean>(null)

  const copyContractsSelected = (): boolean => {
    return selectedOption === 'Copy'
  }

  const addContractsSelected = (): boolean => {
    return selectedOption === 'Add'
  }

  const validateRequest = (): Errors => {
    const newErrors: Errors = {}

    if (!selectedOption) {
      newErrors.selectedOption = 'Please select an option'
    }

    if (copyContractsSelected()) {
      if (!sourcePropertyReference) {
        newErrors.sourcePropertyReference =
          'You must enter a source property reference'
      }
      if (!destinationPropertyReference) {
        newErrors.sourcePropertyReference =
          'You must enter a destination property reference'
      }
    }

    if (addContractsSelected()) {
      if (!destinationPropertyReference) {
        newErrors.destinationPropertyReference =
          'You must enter a destination property reference'
      }
      if (!contractReference) {
        newErrors.contractReference = 'You must enter a contract reference'
      }
    }

    return newErrors
  }

  const handleSubmit = (event: React.SyntheticEvent): void => {
    event.preventDefault()

    if (loading) return

    const newErrors = validateRequest()
    setErrors(newErrors)
    setRequestError(null)

    if (Object.keys(newErrors).length > 0) {
      return
    }

    const sourcePropertyReferenceReq = sourcePropertyReference
      .trim()
      .replaceAll(' ', '')
    const destinationPropertyReferenceReq = destinationPropertyReference
      .trim()
      .replaceAll(' ', '')
    const contractReferenceReq = contractReference.trim().replaceAll(' ', '')

    const body = {
      sourcePropertyReference: sourcePropertyReferenceReq,
      destinationPropertyReference: destinationPropertyReferenceReq,
      contractReference: contractReferenceReq,
      mode: selectedOption,
    }

    const url = `/api/backOffice/sor-contracts`

    setLoading(true)

    frontEndApiRequest({
      method: 'post',
      path: url,
      requestData: body,
      params: null,
      paramsSerializer: null,
    })
      .then(() => {
        setFormSuccess(true)
      })
      .catch((err) => {
        console.error(err)
        setRequestError(err.message)
      })
      .finally(() => {
        setLoading(false)
      })
  }

  return (
    <Layout title="SOR Contract Modification">
      {loading ? (
        <Spinner resource={null} />
      ) : (
        <>
          {formSuccess ? (
            <div>
              <SuccessMessage title="SOR contracts modified successfully!" />
              <p>
                <a
                  className="lbh-link"
                  role="button"
                  onClick={() => setFormSuccess(null)}
                >
                  Change more SOR contracts
                </a>
              </p>
            </div>
          ) : (
            <form onSubmit={handleSubmit}>
              {requestError && (
                <>
                  {/*  @ts-ignore */}
                  <ErrorMessage label={requestError} />
                </>
              )}

              <div>
                {/*  @ts-ignore */}
                <ControlledRadio
                  label="Select workflow type"
                  name="selectedOption"
                  options={radioOptions.map((x) => x.value)}
                  onChange={(event) => setSelectedOption(event.target.value)}
                  selectedOption={selectedOption}
                  error={
                    errors.selectedOption && { message: errors.selectedOption }
                  }
                />
              </div>

              {selectedOption === 'Copy' && (
                <div>
                  {/*  @ts-ignore */}
                  <TextInput
                    label="Source property reference (include leading zeroes)"
                    name="sourceInput"
                    placeholder="eg. 00023400"
                    value={sourcePropertyReference}
                    onChange={(event) =>
                      setSourcePropertyReference(event.target.value)
                    }
                    error={
                      errors.sourcePropertyReference && {
                        message: errors.sourcePropertyReference,
                      }
                    }
                  />
                </div>
              )}

              <div>
                {/*  @ts-ignore */}
                <TextInput
                  label="Destination property reference (include leading zeroes)"
                  name="destInput"
                  placeholder="eg. 00023400"
                  value={destinationPropertyReference}
                  onChange={(event) =>
                    setDestinationPropertyReference(event.target.value)
                  }
                  error={
                    errors.destinationPropertyReference && {
                      message: errors.destinationPropertyReference,
                    }
                  }
                />
              </div>

              {selectedOption === 'Add' && (
                <div>
                  {/*  @ts-ignore */}
                  <TextInput
                    label="Contract reference"
                    name="contractRefInput"
                    placeholder="eg. 001-H01-MAT2"
                    value={contractReference}
                    onChange={(event) =>
                      setContractReference(event.target.value)
                    }
                    error={
                      errors.contractReference && {
                        message: errors.contractReference,
                      }
                    }
                  />
                </div>
              )}

              <div>
                {/*  @ts-ignore */}
                <Button label="Execute" type="submit" />
              </div>
            </form>
          )}
        </>
      )}
    </Layout>
  )
}

export default SORContracts
