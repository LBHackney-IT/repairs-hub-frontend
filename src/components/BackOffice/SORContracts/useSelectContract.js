import { useState, useEffect } from 'react'

import { fetchContractors, fetchContracts } from '../requests'

import useSelectContractor from '../AddSORCodes/useSelectContractor'

const useSelectContract = () => {
  const [contractors, setContractors] = useState(null)
  const { selectedContractor, handleSelectContractor } = useSelectContractor(
    contractors
  )

  const [loadingContracts, setLoadingContracts] = useState(false)
  const [contracts, setContracts] = useState(null)
  const [selectedContract, setSelectedContract] = useState(null)

  const [loadingContractors, setLoadingContractors] = useState(false)

  useEffect(() => {
    // load contracts

    setLoadingContractors(true)

    fetchContractors()
      .then((res) => {
        console.log({ res })
        setContractors(res)
      })
      .catch((err) => {
        console.error(err)
      })
      .finally(() => {
        setLoadingContractors(false)
      })
  }, [])

  useEffect(() => {
    handleContractorChange()
  }, [selectedContractor])

  const handleContractorChange = () => {
    if (selectedContractor === null) {
      setContracts(null)
      setSelectedContract(null)
      return
    }

    setLoadingContracts(true)

    fetchContracts(selectedContractor.contractorReference)
      .then((res) => {
        setContracts(res)
      })
      .finally(() => {
        setLoadingContracts(false)
      })
  }

  return {
    contractors,
    handleSelectContractor,
    selectedContractor,
    contracts,
    selectedContract,
    setSelectedContract,
    loadingContracts,
    loadingContractors
  }
}

export default useSelectContract
