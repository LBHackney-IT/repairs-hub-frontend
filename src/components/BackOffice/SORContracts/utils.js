export const sanitizeInput = value => {
    return value.trim().replaceAll(' ', '')
}

export const saveContractChangesToDatabase = async (data) => {
    await frontEndApiRequest({
        method: 'POST',
        path: `/api/backOffice/sor-contracts`,
        requestData: data,
      })
  }