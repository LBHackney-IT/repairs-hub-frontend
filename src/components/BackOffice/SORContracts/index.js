import Layout from '../Layout'
import { useState } from 'react'
import { TextInput, Button } from '../../Form'
import ConfirmationModal from '../Components/ConfirmationModal'
import ControlledRadio from '../Components/ControlledRadio'
import Spinner from '../../Spinner'
import ErrorMessage from '../../Errors/ErrorMessage'
import SuccessMessage from '../Components/SuccessMessage'
import {
  addContractsSelected,
  copyContractsSelected,
  dataToRequestObject,
  propertyReferencesMatch,
  saveContractChangesToDatabase,
  validatePropertyReference,
} from './utils'

import { DataList } from '../../Form'

const radioOptions = [
  {
    text: 'Copy contracts from another property',
    value: 'Copy',
  },
  {
    text: 'Add explicit property and contract references',
    value: 'Add',
  },
]

import useSelectContract from '../hooks/useSelectContract'

const SORContracts = () => {
  const [loading, setLoading] = useState(false)
  const [requestError, setRequestError] = useState(null)
  const [formSuccess, setFormSuccess] = useState(false)

  const [selectedOption, setSelectedOption] = useState(radioOptions[0].value)

  const [sourcePropertyReference, setSourcePropertyReference] = useState('')
  const [
    destinationPropertyReference,
    setDestinationPropertyReference,
  ] = useState('')
  const [showDialog, setShowDialog] = useState(false)

  const {
    contractors,
    handleSelectContractor,
    selectedContractor,
    contracts,
    selectedContract,
    loadingContracts,
    loadingContractors,
    handleSelectContract,
  } = useSelectContract()

  const [errors, setErrors] = useState({})

  const validateRequest = () => {
    const newErrors = {}

    if (!selectedOption) {
      newErrors.selectedOption = 'Please select an option'
    }

    if (!destinationPropertyReference) {
      newErrors.destinationPropertyReference =
        'You must enter a destination property reference'
    } else if (!validatePropertyReference(destinationPropertyReference)) {
      newErrors.destinationPropertyReference = 'PropertyReference is invalid'
    }

    if (copyContractsSelected(selectedOption)) {
      if (!sourcePropertyReference) {
        newErrors.sourcePropertyReference =
          'You must enter a source property reference'
      } else if (!validatePropertyReference(sourcePropertyReference)) {
        newErrors.sourcePropertyReference = 'PropertyReference is invalid'
      } else if (
        propertyReferencesMatch(
          sourcePropertyReference,
          destinationPropertyReference
        )
      ) {
        newErrors.destinationPropertyReference =
          'The destination property reference cannot match source property reference'
      }
    }

    if (addContractsSelected(selectedOption)) {
      if (!selectedContractor) {
        newErrors.contractor = 'You must select a contractor'
      }

      if (!selectedContract) {
        newErrors.contract = 'You must select a contract'
      }
    }

    return newErrors
  }

  const resetForm = () => {
    setSourcePropertyReference('')
    setDestinationPropertyReference('')
    handleSelectContractor(null)
    setFormSuccess(null)
    setErrors({})
    setRequestError(null)
    setShowDialog(false)
  }

  const renderConfirmationModal = () => {
    if (showDialog) {
      return (
        <ConfirmationModal
          title={'Modify SOR contracts?'}
          showDialog
          setShowDialog={setShowDialog}
          modalText={
            'The operation cannot be undone, please make sure the details entered are correct before proceeding.'
          }
          onSubmit={modifySORContracts}
          yesButtonText={'Modify contracts'}
        />
      )
    }
  }

  const handleSubmit = (e) => {
    e.preventDefault()

    if (loading) return

    const newErrors = validateRequest()
    setErrors(newErrors)
    setRequestError(null)

    if (Object.keys(newErrors).length > 0) return

    setShowDialog(true)
  }

  const modifySORContracts = () => {
    setLoading(true)

    const body = dataToRequestObject(
      sourcePropertyReference,
      destinationPropertyReference,
      selectedContract,
      selectedOption
    )

    saveContractChangesToDatabase(body)
      .then(() => {
        setFormSuccess(true)
      })
      .catch((err) => {
        console.error(err)

        if (err?.response?.data?.message !== null) {
          setRequestError(err?.response?.data?.message)
          return
        }

        setRequestError(err)
      })
      .finally(() => {
        setLoading(false)
        setShowDialog(false)
      })
  }

  return (
    <Layout title="SOR contract modification">
      {loading || loadingContractors ? (
        <Spinner />
      ) : (
        <>
          {formSuccess ? (
            <div>
              <SuccessMessage title="SOR contracts modified successfully!" />
              <p>
                <a className="lbh-link" role="button" onClick={resetForm}>
                  Change more SOR contracts
                </a>
              </p>
            </div>
          ) : (
            <form onSubmit={handleSubmit}>
              {requestError && <ErrorMessage label={requestError} />}

              {renderConfirmationModal()}

              <div>
                <ControlledRadio
                  label="Select workflow type"
                  name="selectedOption"
                  options={radioOptions}
                  onChange={(event) => setSelectedOption(event.target.value)}
                  selectedOption={selectedOption}
                  error={
                    errors.selectedOption && { message: errors.selectedOption }
                  }
                />
              </div>

              {selectedOption === 'Copy' && (
                <div>
                  <TextInput
                    data-test="sourcePropertyReference"
                    label="Source property reference (include leading zeroes)"
                    name="sourceInput"
                    placeholder="eg. 00023400"
                    value={sourcePropertyReference}
                    onChange={(event) =>
                      setSourcePropertyReference(event.target.value)
                    }
                    error={
                      errors.sourcePropertyReference && {
                        message: errors.sourcePropertyReference,
                      }
                    }
                  />
                </div>
              )}

              <div>
                <TextInput
                  data-test="destinationPropertyReference"
                  label="Destination property reference (include leading zeroes)"
                  name="destInput"
                  placeholder="eg. 00023400"
                  value={destinationPropertyReference}
                  onChange={(event) =>
                    setDestinationPropertyReference(event.target.value)
                  }
                  error={
                    errors.destinationPropertyReference && {
                      message: errors.destinationPropertyReference,
                    }
                  }
                />
              </div>

              {selectedOption === 'Add' && (
                <div>
                  <DataList
                    name="contractor"
                    label="Contractor"
                    options={contractors?.map((x) => x.contractorName) || []}
                    onChange={handleSelectContractor}
                    error={errors.contractor && { message: errors.contractor }}
                    value={selectedContractor?.contractorName}
                  />

                  {loadingContracts ? (
                    <Spinner />
                  ) : (
                    <div>
                      <DataList
                        name="contract"
                        label="Contract"
                        options={contracts || []}
                        disabled={selectedContractor === null}
                        onChange={handleSelectContract}
                        error={
                          contracts?.length === 0
                            ? { message: 'No contracts found' }
                            : errors.contract && { message: errors.contract }
                        }
                        value={selectedContract}
                      />
                    </div>
                  )}
                </div>
              )}

              <div>
                <Button
                  data-test="submit-button"
                  label="Save changes"
                  type="submit"
                />
              </div>
            </form>
          )}
        </>
      )}
    </Layout>
  )
}

export default SORContracts
