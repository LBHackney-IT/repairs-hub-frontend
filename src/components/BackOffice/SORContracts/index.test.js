const mockUseQueryResult = {
  data: [],
  isLoading: false,
  error: null,
}

jest.mock('react-query', () => ({
  QueryClient: jest.fn(),
  QueryClientProvider: ({ children }) => children,
  useQuery: () => mockUseQueryResult,
}))
import { render } from '@testing-library/react'
import SORContracts from './index'

describe('SORContracts component', () => {
  it('should render the component', async () => {
    const { asFragment } = render(<SORContracts />)

    expect(asFragment()).toMatchSnapshot()
  })
})
