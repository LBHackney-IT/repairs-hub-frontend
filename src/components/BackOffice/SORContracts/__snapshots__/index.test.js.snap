// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SORContracts component should render the component 1`] = `
<DocumentFragment>
  <a
    class="govuk-back-link lbh-back-link govuk-!-display-none-print"
    role="button"
  >
    Back
  </a>
  <h1
    class="lbh-heading-h1"
  >
    SOR Contract Modification
  </h1>
  <form>
    <div>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <label
          class="govuk-label govuk-label--m"
          for="selectedOption"
        >
          Select workflow type 
        </label>
        <div
          class="govuk-radios lbh-radios"
        >
          <div
            class="govuk-radios__item"
          >
            <input
              checked=""
              class="govuk-radios__input"
              data-testid="selectedOption"
              id="selectedOption_Copy"
              name="selectedOption"
              type="radio"
              value="Copy"
            />
            <label
              class="govuk-label lbh-label govuk-radios__label"
              for="selectedOption_Copy"
            >
              Copy contracts from another property
            </label>
          </div>
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input"
              data-testid="selectedOption"
              id="selectedOption_Add"
              name="selectedOption"
              type="radio"
              value="Add"
            />
            <label
              class="govuk-label lbh-label govuk-radios__label"
              for="selectedOption_Add"
            >
              Add explicit property and contract references
            </label>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="govuk-form-group lbh-form-group"
        id="sourceInput-form-group"
      >
        <label
          class="govuk-label lbh-label"
          for="sourceInput"
        >
          Source property reference (include leading zeroes) 
        </label>
        <input
          class="govuk-input lbh-input undefined"
          data-testid="sourceInput"
          id="sourceInput"
          name="sourceInput"
          placeholder="eg. 00023400"
          type="text"
          value=""
        />
      </div>
    </div>
    <div>
      <div
        class="govuk-form-group lbh-form-group"
        id="destInput-form-group"
      >
        <label
          class="govuk-label lbh-label"
          for="destInput"
        >
          Destination property reference (include leading zeroes) 
        </label>
        <input
          class="govuk-input lbh-input undefined"
          data-testid="destInput"
          id="destInput"
          name="destInput"
          placeholder="eg. 00023400"
          type="text"
          value=""
        />
      </div>
    </div>
    <div>
      <button
        class="govuk-button lbh-button"
        data-module="govuk-button"
        type="submit"
      >
        Execute
      </button>
    </div>
  </form>
</DocumentFragment>
`;
